components/script/dom/textencoderstream.rs

```diff
@@ -0,0 +1,443 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::borrow::Cow;
+use std::cell::Cell;
+use std::ptr;
+use std::rc::Rc;
+
+use dom_struct::dom_struct;
+use js::conversions::{
+    ConversionResult, FromJSValConvertible, ToJSValConvertible, latin1_to_string,
+};
+use js::jsapi::{
+    IsArrayObject, JS_DeprecatedStringHasLatin1Chars, JS_GetTwoByteStringCharsAndLength, JSObject,
+    JSType, ToPrimitive,
+};
+use js::jsval::{JSVal, UndefinedValue};
+use js::rust::{HandleObject as SafeHandleObject, HandleValue as SafeHandleValue, IntoHandle};
+use js::typedarray::Uint8Array;
+
+use crate::dom::bindings::buffer_source::create_buffer_source;
+use crate::dom::bindings::codegen::Bindings::TextEncoderStreamBinding::TextEncoderStreamMethods;
+use crate::dom::bindings::error::{Error, Fallible};
+use crate::dom::bindings::reflector::{Reflector, reflect_dom_object_with_proto};
+use crate::dom::bindings::root::{Dom, DomRoot};
+use crate::dom::bindings::str::DOMString;
+use crate::dom::transformstreamdefaultcontroller::TransformerType;
+use crate::dom::types::{GlobalScope, TransformStream, TransformStreamDefaultController};
+use crate::script_runtime::{CanGc, JSContext as SafeJSContext};
+use crate::{DomTypeHolder, DomTypes};
+
+/// String converted from an input JS Value
+enum ConvertedInput<'a> {
+    Str(&'static str),
+    String(String),
+    CodeUnits(&'a [u16]),
+}
+
+/// Converts a JSVal to a potentially ill-formed string. This is re-implemented
+/// instead of using `js::rust::ToString` because the latter shows unexpected
+/// behavior when handling conversion of non-string JS types into strings
+///
+/// <https://tc39.es/ecma262/multipage/abstract-operations.html#sec-tostring>
+#[allow(unsafe_code)]
+fn jsval_to_string<'a>(
+    cx: SafeJSContext,
+    value: SafeHandleValue, // TODO: how to argure the lifetime here
+) -> Fallible<ConvertedInput<'a>> {
+    // Step 1. If argument is a String, return argument.
+    if value.is_string() {
+        let jsstr = std::ptr::NonNull::new(value.to_string())
+            .ok_or_else(|| Error::Type("Failed to convert to JSString".to_owned()))?;
+        unsafe {
+            if JS_DeprecatedStringHasLatin1Chars(jsstr.as_ptr()) {
```

Comment:
not now, but need to hide these unsafe functions in future inside a safe wrapper.